File: 0438-find-all-anagrams-in-a-string/0438-find-all-anagrams-in-a-string.py
class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:

        res = []        

        if len(s) < len(p):
            return []

        sCount, pCount = {}, {}

        for i in range(len(p)):
            pCount[p[i]] = pCount.get(p[i], 0) + 1
            sCount[s[i]] = sCount.get(s[i], 0) + 1

        if pCount == sCount:
            res.append(0);

        left = 0
        for right in range(len(p), len(s)):
            sCount[s[right]] = sCount.get(s[right], 0) + 1
            sCount[s[left]] -= 1
            if sCount[s[left]] == 0:
                del sCount[s[left]]

            left += 1                

            if sCount == pCount:
                res.append(left)


        return res

File: 0438-find-all-anagrams-in-a-string/0438-find-all-anagrams-in-a-string.java
class Solution {
    public List<Integer> findAnagrams(String s, String p) {

        List<Integer> res = new ArrayList<Integer>();

        if(p.length() > s.length()) return res;

        int arr[] = new int[26];      

        for(char c :  p.toCharArray()){
            arr[c - 'a']++;
        }

        int left = 0;
        int count = p.length(); 

        for(int right=0; right<s.length(); right++){
            if(arr[s.charAt(right) - 'a'] > 0){
                count--;
            } 

            arr[s.charAt(right) - 'a']--;

            if(count == 0) res.add(left);

            if(right-left+1 == p.length()){
                if(arr[s.charAt(left) - 'a'] >= 0) count++;

                arr[s.charAt(left) - 'a']++;

                left++;
            }
        }

        return res;
    }
}